src/main/scala/cromwell/engine/WorkflowDescriptor.scala

```diff
@@ -0,0 +1,196 @@
+    if (optionCacheReading) logReadDisabled()
+  }
+
+  lazy val writeToCache = configCallCaching && optionCacheWriting
+  lazy val readFromCache = configCallCaching && optionCacheReading
+
+  lazy val workflowLogger = props.get("LOG_MODE") match {
+    case Some(x) if x.toUpperCase.contains("SERVER") => makeFileLogger(
+      Paths.get(props.getOrElse("LOG_ROOT", ".")),
+      s"workflow.$id.log",
+      Level.toLevel(props.getOrElse("LOG_LEVEL", "debug"))
+    )
+    case _ => NOPLogger.NOP_LOGGER
+  }
+
+  private def makeFileLogger(root: Path, name: String, level: Level): Logger = {
+    val ctx = LoggerFactory.getILoggerFactory.asInstanceOf[LoggerContext]
+    val encoder = new PatternLayoutEncoder()
+    encoder.setPattern("%date %-5level - %msg%n")
+    encoder.setContext(ctx)
+    encoder.start()
+
+    val path = root.resolve(name).toAbsolutePath.toString
+    val appender = new FileAppender[ILoggingEvent]()
+    appender.setFile(path)
+    appender.setEncoder(encoder)
+    appender.setName(name)
+    appender.setContext(ctx)
+    appender.start()
+
+    val fileLogger = ctx.getLogger(name)
+    fileLogger.addAppender(appender)
+    fileLogger.setAdditive(false)
+    fileLogger.setLevel(level)
+    fileLogger
+  }
+
+  private def logWriteDisabled() = workflowLogger.warn(writeDisabled)
+  private def logReadDisabled() = workflowLogger.warn(readDisabled)
+}
+
+object WorkflowDescriptor {
+  def apply(id: WorkflowId, sourceFiles: WorkflowSourceFiles): WorkflowDescriptor = {
+    WorkflowDescriptor(id, sourceFiles, ConfigFactory.load)
+  }
+
+  def apply(id: WorkflowId, sourceFiles: WorkflowSourceFiles, conf: Config): WorkflowDescriptor = {
+    validateWorkflowDescriptor(id, sourceFiles, CromwellServer.backend, conf) match {
+      case scalaz.Success(w) => w
+      case scalaz.Failure(f) =>
+        throw new IllegalArgumentException(s"""Workflow ${id.toString} failed to process inputs:\n${f.list.mkString("\n")}""")
+    }
+  }
+
+  private def validateWorkflowDescriptor(id: WorkflowId,
+                                         sourceFiles: WorkflowSourceFiles,
+                                         backend: Backend,
+                                         conf: Config): ErrorOr[WorkflowDescriptor] = {
+    val namespace = validateNamespace(sourceFiles.wdlSource, backend)
+    val rawInputs = validateRawInputs(id, sourceFiles.inputsJson)
+    val options = validateWorkflowOptions(id, sourceFiles.workflowOptionsJson, backend)
+    (namespace |@| rawInputs |@| options) { (_, _, _) } match {
+      case scalaz.Success((n, r, o)) =>
+        val validatedDescriptor = for {
+          c <- validateCoercedInputs(id, r, n).disjunction
+          d <- validateDeclarations(id, n, o, c, backend).disjunction
+        } yield WorkflowDescriptor(id, sourceFiles, o, r, n, c, d, backend, configCallCaching(conf), lookupDockerHash(conf))
+        validatedDescriptor.validation
+      case scalaz.Failure(f) => f.list.mkString("").failureNel
```

Comment:
I doubt we have a test for multiple failures? Maybe in this PR, as this is copy paste, we don't need one. But I also doubt the failure strings are supposed to be concatenated without at least a `", "`.